distributed/worker.py

```diff
@@ -1406,6 +1409,11 @@ async def close(
                 logger.info("Closed worker has not yet started: %s", self.status)
             self.status = Status.closing
 
+            for task in self._async_instructions:
+                task.cancel()
+            while self._async_instructions:
+                await asyncio.sleep(0)
```

Comment:
probably will need an `asyncio.gather(*self._async_instructions, return_exceptions=True)` and something to handle close itself being cancelled

Reply:
if close itself is cancelled, it should not wait for other tasks being cancelled to return (which will happen at the next event loop cycle anyway)

Reply:
Tasks can handle the cancel eg:


```python
async def example():
    async with database():
        await asyncio.Event().wait()
```

The worker will need to wait for the async context manager to close

Reply:
First, this isn't a plugin system. We don't have any tasks there that catch CancelledError.
Second,
```python
            while self._async_instructions:
                await asyncio.sleep(0)
``` 

is doing exactly the same thing as the asyncio.gather you suggested.

Reply:
a user can submit an async function via `c.submit(f, ev, key="f1")` which is likely to suppress cancellation. using `while self._async_instructions: await asyncio.sleep(0)` burns the cpu while waiting for cancellation to be handled

Reply:
See latest commit